github: limit test runs if only Go test files have changed

jrajahalme opened the pull request:
Skip workflows that do not run Go tests if only Go test files have changed.

Glob syntax reference: https://github.com/dorny/paths-filter/issues/97

jrajahalme committed:
DO NOT MERGE

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

pchaigno commented on .github/workflows/conformance-aks-v1.11.yaml:
```diff
@@ -112,10 +112,11 @@ jobs:
           filters: |
             src:
-              - '!(test|Documentation)/**'
+              - '!(test|Documentation)/**/!(*._test.go)'
```
I think you want:
```suggestion
              - '!(test|Documentation)/**|!(*._test.go)'
```
That said, I'm unsure it will work because NOT A OR NOT B isn't the same as NOT (A OR B) and the later is what you seem to be after. I'm unsure if it can be written as a filter expression here.

jrajahalme replied:
Did you look at https://github.com/dorny/paths-filter/issues/97?

jrajahalme replied:
Before we had
- "any path (`/**`) that starts with a directory that is not `test` or `Documentation` (`!(test|Documentation)`)
plan is to make it 
- "any path not ending in `_test.go`(`/**/!(*_test.go)`) that starts with a directory that is not `test` or `Documentation` (`!(test|Documentation)`)

I see I have an extra `.` in there right before `_test.go`, will remove.

pchaigno replied:
Ah, thanks for the pointer and explanation. That GitHub Action has got more advanced filters than I remembered.

What I'm a bit :suspicious: about is the double negation. My understanding was that a `!` negates everything after it. Hence, the current filter would also match files at the root of the repository. If that is indeed the case, then does the first `!` negate the second?

(Maybe it's worth adding a test case for this in the upstream repository as suggested by the author: https://github.com/dorny/paths-filter/blob/4067d885736b84de7c414f582ac45897079b0a78/__tests__/filter.test.ts#L101.)

jrajahalme replied:
Added test: https://github.com/dorny/paths-filter/pull/191
Local run (`matches all except tests or docs and tsx or less files (multiple negate groups)` is the new one):
```
$ ./node_modules/.bin/jest __tests__/filter.test.ts --config jest.config.js 
 PASS  __tests__/filter.test.ts
  yaml filter parsing tests
    ✓ throws if yaml is not a dictionary (12 ms)
    ✓ throws if pattern is not a string (2 ms)
  matching tests
    ✓ matches single inline rule (1 ms)
    ✓ matches single rule in single group
    ✓ no match when file is in different folder
    ✓ match only within second groups 
    ✓ match only withing second rule of single group
    ✓ matches anything
    ✓ globbing matches path where file or folder name starts with dot (1 ms)
    ✓ matches all except tsx and less files (negate a group with or-ed parts)
    ✓ matches all except tests or docs and tsx or less files (multiple negate groups)
    ✓ matches path based on rules included using YAML anchor (1 ms)
  matching specific change status
    ✓ does not match modified file as added
    ✓ match added file as added (1 ms)
    ✓ matches when multiple statuses are configured
    ✓ matches when using an anchor

Test Suites: 1 passed, 1 total
Tests:       16 passed, 16 total
Snapshots:   0 total
Time:        1.741 s, estimated 2 s
Ran all test suites matching /__tests__\/filter.test.ts/i.
```

pchaigno replied:
Nice!

jrajahalme committed:
github: limit test runs if only Go test files have changed

Skip workflows that do not run Go tests if only Go test files have
changed.

Glob syntax reference: https://github.com/dorny/paths-filter/issues/97

The new glob spec should be read as:
- Any path stat starts with a directory that is not 'test' or
  'Documentation' that does not end in '_test.go'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
DO NOT MERGE

Changed 'ci-gke' to only run on pull request if files outside of
'.github', 'test', and 'Documentation', and that are not Go test files.

See if the test run of 'ci-gke' is skipped, while e.g., 'ci-eks' is not.

Author: Jarno Rajahalme <jarno@isovalent.com>
Date:   2023-05-12 09:46:37 +0300

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
